benchmarks/transformer/sdpa.py

```diff
@@ -0,0 +1,175 @@
+    config: ExperimentConfig
+    results: ExperimentResults
+
+    def asdict(self):
+        dict1 = asdict(self.config)
+        dict2 = asdict(self.results)
+        return {**dict1, **dict2}
+
+
+def get_input(
+    config: ExperimentConfig,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    q = torch.randn(
+        (config.batch_size, config.num_heads, config.q_seq_len, config.head_dim),
+        dtype=config.dtype,
+        device=config.device,
+        requires_grad=True,
+    )
+    k = torch.randn(
+        (config.batch_size, config.num_heads, config.kv_seq_len, config.head_dim),
+        dtype=config.dtype,
+        device=config.device,
+        requires_grad=True,
+    )
+    v = torch.randn(
+        (config.batch_size, config.num_heads, config.kv_seq_len, config.head_dim),
+        dtype=config.dtype,
+        device=config.device,
+        requires_grad=True,
+    )
+    return q, k, v
+
+
+def run_single_experiment(config: ExperimentConfig) -> ExperimentResults:
+    q, k, v = get_input(config)
+    is_causal = config.is_causal
+
+    forward_time = benchmark_torch_function_in_microseconds(
+        scaled_dot_product_attention,
+        q,
+        k,
+        v,
+        is_causal=is_causal,
+        attn_mask=None,
+    )
+    out_torch = scaled_dot_product_attention(
+        q, k, v, is_causal=is_causal, attn_mask=None
+    )
+    dOut = torch.randn_like(out_torch)
+    backward_time = benchmark_torch_function_in_microseconds(
+        out_torch.backward, dOut, retain_graph=True
+    )
+
+    return ExperimentResults(
+        forward_time=forward_time,
+        backward_time=backward_time,
+    )
+
+
+def generate_experiment_configs() -> List[ExperimentConfig]:
+    batch_sizes = [
+        1,
+        8,
+    ]
+    num_heads = [16]
+    q_kv_seq_lens = [(128, 128), (256, 256), (512, 512), (1024, 1024)]
+    embed_dims = [2048]
+    dtypes = [
+        torch.bfloat16,
+    ]
+    is_causal = [True, False]
+    all_configs = []
+    for (
+        bsz,
+        heads,
+        (q_seq_len, kv_seq_len),
+        embed_dim,
+        causal,
+        dtype,
+    ) in itertools.product(
+        batch_sizes, num_heads, q_kv_seq_lens, embed_dims, is_causal, dtypes
+    ):
+        all_configs.append(
+            ExperimentConfig(
+                batch_size=bsz,
+                num_heads=heads,
+                q_seq_len=q_seq_len,
+                kv_seq_len=kv_seq_len,
+                embed_dim=embed_dim,
+                is_causal=causal,
+                dtype=dtype,
+            )
+        )
+
+    return all_configs
+
+
+def print_results(experiments: List[Experiment]):
+    table_data = defaultdict(list)
+    for experiment in experiments:
+        for key, value in experiment.asdict().items():
+            table_data[key].append(value)
+    del table_data["device"]
+    print(tabulate(table_data, headers="keys", tablefmt="pretty", floatfmt=".3f"))
+
+
+def main():
+    seed = 123
+    np.random.seed(seed)
```

Comment:
nit: maybe missing something but why do we need to set numpy seed when no np functions are called?

Reply:
yeah this was belt and suspenders, will remove